erpnext/selling/doctype/sales_order/test_sales_order.py

```diff
@@ -295,6 +295,45 @@ def test_auto_insert_price(self):
 			{"price_list": "_Test Price List", "item_code": "_Test Item for Auto Price List"}, "price_list_rate"), None)
 
 		frappe.db.set_value("Stock Settings", None, "auto_insert_price_list_rate_if_missing", 1)
+	
+	def test_drop_shipping(self):
+		from erpnext.selling.doctype.sales_order.sales_order import make_drop_shipment, make_delivery_note
+		from erpnext.stock.doctype.item.test_item import make_item
+
+		po_item = make_item("_Test Item for Drop Shipping", {"is_stock_item": 1, "is_sales_item": 1,
+			"is_purchase_item": 1, "is_drop_ship": 1, 'default_supplier': '_Test Supplier'})
+
+		dn_item = make_item("_Test Regular Item", {"is_stock_item": 1, "is_sales_item": 1,
+			"is_purchase_item": 1})
+
+		so_items = [
+			{
+				"item_code": po_item.item_code,
+				"warehouse": "_Test Warehouse - _TC",
+				"qty": 1,
+				"rate": 400,
+				"conversion_factor": 1.0,
+				"is_drop_ship": 1,
+				"supplier": '_Test Supplier'
+			},
+			{
+				"item_code": dn_item.item_code,
+				"warehouse": "_Test Warehouse - _TC",
+				"qty": 1,
+				"rate": 300,
+				"conversion_factor": 1.0
+			}
+		]
+
+		so = make_sales_order(item_list=so_items)
+		po = make_drop_shipment(so.name, '_Test Supplier')
+		dn = make_delivery_note(so.name)
+
+		self.assertEquals(so.customer, po.customer)
+		self.assertEquals(po.items[0].prevdoc_doctype, "Sales Order")
+		self.assertEquals(po.items[0].prevdoc_docname, so.name)
+		self.assertEquals(po.items[0].item_code, po_item.item_code)
+		self.assertEquals(dn.items[0].item_code, dn_item.item_code)
```

Comment:
- Test cases to check ordered_qty and per_ordered
- Check length of po_items
- Logic of Closed status with reserved and ordered qty